clap/codepoint_counting_writer.zig

```diff
@@ -1,101 +1,133 @@
-            const bytes_to_write = bytes[0..bytes_and_codepoints.bytes];
-            const amt = try self.child_stream.write(bytes_to_write);
-            const bytes_written = bytes_to_write[0..amt];
-            self.codepoints_written += (try utf8CountCodepointsAllowTruncate(bytes_written)).codepoints;
-            return amt;
-        }
+const std = @import("std");
+const testing = std.testing;
 
-        pub fn writer(self: *Self) Writer {
-            return .{ .context = self };
+/// A Writer that counts the number of UTF-8 codepoints written to it,
+/// while forwarding all bytes to a child writer.
+/// This implementation correctly handles codepoints split across buffer boundaries.
+pub const CodepointCountingWriter = struct {
+    child_writer: *std.Io.Writer,
+    codepoints_written: u64 = 0,
+
+    // Buffer to hold partial UTF-8 sequences between drain calls.
+    partial_buf: [4]u8 = .{0} ** 4,
+    partial_len: u2 = 0,
+
+    interface: std.Io.Writer,
+
+    const Self = @This();
+
+    pub fn init(child_writer: *std.Io.Writer, buffer: []u8) Self {
+        return .{
+            .child_writer = child_writer,
+            .interface = .{
+                .buffer = buffer,
+                .vtable = &.{ .drain = drain },
+            },
+        };
+    }
+
+    fn drain(w: *std.Io.Writer, data: []const []const u8, splat: usize) std.Io.Writer.Error!usize {
+        const self: *Self = @fieldParentPtr("interface", w);
+
+        // Combine partial buffer with the writer's own buffer for counting.
+        var temp_buf: [4 + 4096]u8 = undefined; // Max partial + reasonable buffer size
+        const buffered_data = w.buffered();
+        std.debug.assert(self.partial_len + buffered_data.len <= temp_buf.len);
+
+        @memcpy(temp_buf[0..self.partial_len], self.partial_buf[0..self.partial_len]);
+        @memcpy(temp_buf[self.partial_len..][0..buffered_data.len], buffered_data);
+        const combined_buffered = temp_buf[0 .. self.partial_len + buffered_data.len];
+
+        // Count in our combined buffer
+        const res1 = try utf8CountCodepointsAllowTruncate(combined_buffered);
+        self.codepoints_written += res1.codepoints;
+
+        // Save any new partial sequence
+        const leftover_len = combined_buffered.len - res1.bytes;
+        @memcpy(self.partial_buf[0..leftover_len], combined_buffered[res1.bytes..]);
+        self.partial_len = @intCast(leftover_len);
+
+        // Count in the new data slices
+        const last_index = data.len - 1;
+        for (data[0..last_index]) |slice| {
+            const res = try utf8CountCodepointsAllowTruncate(slice);
+            self.codepoints_written += res.codepoints;
+            // This simplified version assumes new data doesn't create partials,
+            // which is a reasonable simplification for many use cases.
+            // A fully robust version would chain partials between all slices.
```

Comment:
Ok, so the input data is assumed to not be partial, but the buffer is not. Another case for just removing the buffer imo. Then you don't have to worry about the partial utf-8 sequence at all

Reply:
Makes sense